File: packages/paygate-python/paygate/cli/doctor.py
"""`paygate doctor` — check config + connectivity."""

from __future__ import annotations

import asyncio
import os

import httpx
import typer

from paygate.config import load_config
from paygate.constants import DEFAULT_FACILITATOR_URL

doctor_app = typer.Typer(no_args_is_help=False, invoke_without_command=True)


@doctor_app.callback(invoke_without_command=True)
def run(
    config: str = typer.Option("./paygate.config.yml", "-c", "--config"),
) -> None:
    results: list[tuple[str, bool, str]] = []
    try:
        cfg = load_config(config)
        results.append(("config loaded", True, ""))
    except Exception as e:
        results.append(("config loaded", False, str(e)))
        cfg = None  # type: ignore[assignment]

    if cfg is not None:
        if cfg.wallets.base:
            ok, detail = asyncio.run(_rpc_ok(os.environ.get("PAYGATE_BASE_RPC_URL", "https://mainnet.base.org"), "evm"))
            results.append(("base rpc", ok, detail))
        if cfg.wallets.solana:
            ok, detail = asyncio.run(_rpc_ok(os.environ.get("PAYGATE_SOLANA_RPC_URL", "https://api.mainnet-beta.solana.com"), "solana"))
            results.append(("solana rpc", ok, detail))

        if cfg.defaults.facilitator == "coinbase":
            ok, detail = asyncio.run(
                _facilitator_ok(cfg.advanced.facilitator_url or DEFAULT_FACILITATOR_URL)
            )
            results.append(("facilitator", ok, detail))

    pad = max(len(r[0]) for r in results) if results else 0
    ok_all = True
    for name, ok, detail in results:
        mark = "[ OK ]" if ok else "[FAIL]"
        typer.echo(f"{mark}  {name.ljust(pad)}  {detail}".rstrip())
        if not ok:
            ok_all = False
    raise typer.Exit(code=0 if ok_all else 1)


async def _rpc_ok(url: str, kind: str) -> tuple[bool, str]:
    try:
        async with httpx.AsyncClient(timeout=3) as client:
            if kind == "evm":
                r = await client.post(url, json={"jsonrpc": "2.0", "id": 1, "method": "eth_blockNumber"})
            else:
                r = await client.post(url, json={"jsonrpc": "2.0", "id": 1, "method": "getSlot"})
            if r.status_code >= 400:
                return False, f"HTTP {r.status_code}"
            return True, ""
    except Exception as e:
        return False, str(e)


async def _facilitator_ok(url: str) -> tuple[bool, str]:
    try:
        async with httpx.AsyncClient(timeout=3) as client:
            r = await client.get(url.rstrip("/") + "/health")
            return r.status_code < 500, ""
    except Exception as e:
        return False, str(e)

File: packages/paygate-node/src/cli/doctor.ts
import Redis from 'ioredis';
import { loadConfigFromFile } from '../config.js';
import { DEFAULT_FACILITATOR_URL, USDC_ADDRESSES } from '../constants.js';
import { createPublicClient, http } from 'viem';
import { base, baseSepolia } from 'viem/chains';
import { Connection } from '@solana/web3.js';
import { FacilitatorClient } from '../facilitator/client.js';

export interface DoctorOptions {
  readonly config: string;
}

export async function runDoctor(opts: DoctorOptions): Promise<boolean> {
  const results: Array<{ name: string; ok: boolean; detail?: string }> = [];

  try {
    const cfg = loadConfigFromFile(opts.config);
    results.push({ name: 'config loaded', ok: true });

    if (cfg.wallets.base) {
      const ok = await rpcReachable(
        process.env['PAYGATE_BASE_RPC_URL'] ?? 'https://mainnet.base.org',
        'evm',
      );
      results.push({ name: 'base rpc', ok });
    }
    if (cfg.wallets.solana) {
      const ok = await rpcReachable(
        process.env['PAYGATE_SOLANA_RPC_URL'] ?? 'https://api.mainnet-beta.solana.com',
        'solana',
      );
      results.push({ name: 'solana rpc', ok });
    }

    if (process.env['PAYGATE_REDIS_URL']) {
      const ok = await redisReachable(process.env['PAYGATE_REDIS_URL']);
      results.push({ name: 'redis', ok });
    }

    if (cfg.defaults.facilitator === 'coinbase') {
      const client = new FacilitatorClient({
        url: cfg.advanced.facilitator_url ?? DEFAULT_FACILITATOR_URL,
      });
      results.push({ name: 'facilitator', ok: await client.health() });
    }
  } catch (err) {
    results.push({ name: 'config loaded', ok: false, detail: (err as Error).message });
  }

  const pad = Math.max(...results.map((r) => r.name.length));
  let allOk = true;
  for (const r of results) {
    const mark = r.ok ? '[ OK ]' : '[FAIL]';
    const row = `${mark}  ${r.name.padEnd(pad)}  ${r.detail ?? ''}`.trimEnd();
    console.log(row);
    if (!r.ok) allOk = false;
  }
  return allOk;
}

async function rpcReachable(url: string, kind: 'evm' | 'solana'): Promise<boolean> {
  try {
    if (kind === 'evm') {
      const client = createPublicClient({
        chain: url.includes('sepolia') ? baseSepolia : base,
        transport: http(url),
      });
      await client.getBlockNumber();
    } else {
      const conn = new Connection(url, 'confirmed');
      await conn.getSlot();
    }
    return true;
  } catch {
    return false;
  }
}

async function redisReachable(url: string): Promise<boolean> {
  const redis = new Redis(url, { lazyConnect: true, maxRetriesPerRequest: 1 });
  try {
    await redis.connect();
    const pong = await redis.ping();
    await redis.quit();
    return pong === 'PONG';
  } catch {
    return false;
  }
}

void USDC_ADDRESSES;
